python/nav/ipdevpoll/shadows/__init__.py

```diff
@@ -259,6 +267,42 @@ def _fix_binary_garbage(self):
                 setattr(self, attr, repr(value))
         self.clear_cached_objects()
 
+    def _detect_version_changes(self):
+        """
+        Detects if the software, hardware or firmware version changed for each device.
+
+        Saves this information in changed_versions in the Device instance.
+        """
+        old_device = self.get_existing_model()
+        self.changed_versions = {}
+        if old_device:
+            changed_versions = set(self.get_diff_attrs(old_device)).intersection(
+                (
+                    'hardware_version',
+                    'software_version',
+                    'firmware_version',
+                )
+            )
+            for version in changed_versions:
+                self.changed_versions[version] = getattr(old_device, version)
+
+    def cleanup(self, containers):
+        if self.changed_versions:
+            self._post_events_version_changes()
+
+    def _post_events_version_changes(self):
+        """Posts events for software, hardware or firmware changes."""
+        device = self.get_existing_model()
+        for alert_type, old_version in self.changed_versions.items():
+            self.event.notify(
+                device=device,
+                subid=device.id,
```

Comment:
`subid` should be a reference to a sub-entity of a `Netbox` (e.g. for `moduleDown` alerts, `subid` should normally be the primary key of the related `Module` object). I'm thinking two things here:

1. This alert doesn't even identify the current netbox. It should. I.e. the event's `netbox` attribute should also be set.
2. Since this is otherwise an event about a 'generic' Device entry, I don't think you should need to add a `subid` attribute at all (atm. you're adding `device.id`, but you already added a `device` attribute, so it's superfluous)